accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,376 @@
+
+Due to the performance and usability issues, workarounds and alternatives are used including:
+- [MethodInfo.CreateDelegate()](https://learn.microsoft.com/dotnet/api/system.reflection.methodinfo.createdelegate) or [`Delegate.CreateDelegate()`](https://learn.microsoft.com/dotnet/api/system.delegate.createdelegate) which supports a direct, fast method invocation. However, since delegates are strongly-typed, this approach does not work for loosely-typed invoke scenarios where the signature is not known at compile-time.
+- [Dynamic methods](https://learn.microsoft.com/dotnet/framework/reflection-and-codedom/how-to-define-and-execute-dynamic-methods) are used which are IL-emit based and require a non-trivial implementation for even simple things like setting property values. Those who use dynamic methods must have their own loosely-typed invoke APIs, which may go as far as using generics with `Type.MakeGenericType()` or `MethodInfo.MakeGenericMethod()` to avoid boxing. In addition, a fallback to standard reflection is required to support those platforms where IL Emit is not available.
+- Compiled [expression trees](https://learn.microsoft.com/dotnet/csharp/programming-guide/concepts/expression-trees/) which use dynamic methods if IL Emit is available but also conveniently fallback to standard reflection when not available. Using an expression to invoke a member isn't intuitive and brings along the large `System.Linq.Expressions.dll` assembly.
+
+# .NET 8 Goals
+The .NET 7 release had a 3-4x perf improvement for the existing `Invoke()` APIs by using IL Emit when available and falling back to standard reflection when IL Emit is not available. Although this improvement is significant, it still doesn't replace the need to use the IL-Emit based alternatives (dynamic methods and expression trees) for highly-performance-sensitive scenarios including the `System.Text.Json` serializer. New APIs are required that don't have the overhead of the existing `Invoke()` APIs.
+
+For .NET 8, there are two primary goals:
+1) Support byref-like types both for invoking and passing as arguments; this unblocks various scenarios. An unsafe approach may used for .NET 8 if support for `TypedReference` isn't addressed by Roslyn (covered later).
+2) Support "fast invoke" so that dynamic methods no longer have to be used for performance. Today both `STJ (System.Text.Json)` and [DI (dependency injection)](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection) use IL Emit for performance although DI uses emit through expressions while STJ uses IL Emit directly.
+
+## STJ and DI
+As a litmus test, STJ and DI will be changed to use the new APIs proposed here. This is more important to DI since, unlike STJ which has a source generator that can avoid reflection, DI must continue to use reflection. See also https://github.com/dotnet/runtime/issues/66153 which should be addressed by having a fast constructor invoke that can be used by DI.
```

Comment:
That is the thinking - if it's not fast enough for internal use, it may not be fast enough for others. However, I have a prototype for STJ using the proposed APIs and has a 5% regression in deserialization (note that without any Emit, deserialization is ~2x slower, so getting it to just 5% is pretty significant). That was the reason I also added the `PropertyInfo.CreateGetter\SetterDelegate` to match how STJ uses delegates today. There may also be some gains in going down an unsafe approach, but I haven't measured the perf of that yet.

For DI, I hope to use the new APIs directly.